### PR TITLE
feat(ext/node): add AES-CCM cipher support

### PR DESCRIPTION
**Pull request by bartlomieju**
## Summary

- Implements full AES-CCM (RFC 3610) authenticated encryption for 128/192/256-bit keys
- Adds `AesCcmCipher` struct with CBC-MAC + CTR mode per RFC 3610
- Supports CCM-specific `setAAD` with `plaintextLength` option (required by Node.js API)
- Buffers data in `update()` and processes in `final()` since CCM requires knowing plaintext length upfront for CBC-MAC
- Validates nonce length (7-13 bytes), tag length (4-16 even), and message size limits

This is the last extraction from #32745 -- sign/verify (#33083), GCM (#33079), and ChaCha20-Poly1305 (#33084) are already landed.

## Test plan

- [x] `./x test-compat parallel/test-crypto-authenticated.js` passes
- [x] `./x test-node crypto_cipher` passes (existing GCM + cipher tests)
- [x] `cargo clippy -p deno_node_crypto` clean
- [x] `./tools/format.js` clean
- [x] `./tools/lint.js --js` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by kajukitli** [comment]
security review pass:

• i don't see any new non-constant-time auth tag comparisons in the rust side. the tag checks that are in scope are using `subtle::ConstantTimeEq` (`ct_eq`) for gcm/chacha20-poly1305, so nothing jumped out there.

• i *did* find a blocking correctness/security issue though: the PR wires a lot of AES-CCM behavior into the js polyfill, but the rust backend ops don't appear to implement CCM at all.
  • `ext/node_crypto/lib.rs` still exposes `op_node_cipheriv_set_aad` / `op_node_decipheriv_set_aad` with only `(rid, aad)` — the new js code passes a `plaintextLength`, but it never reaches rust.
  • `ext/node_crypto/cipher.rs` doesn't appear to add any `aes-*-ccm` variants in `Cipher` / `Decipher`, constructor matching, or final/auth logic.
  • so this currently looks like the js layer claims CCM semantics, but the crypto backend can't actually enforce/process them.

for an AEAD mode that's a problem, because `plaintextLength` is part of the CCM contract and the backend needs to own the MAC/encrypt/decrypt behavior. otherwise best case this is just broken, worst case the API surface suggests authenticated encryption support that isn't actually implemented correctly.

so: constant-time-wise this looks fine, but i don't think the PR is ready yet until CCM is implemented end-to-end in the backend and the js `setAAD(..., { plaintextLength })` path is actually plumbed through.

**Comment by bartlomieju** [comment]
Thanks for the security review! The constant-time check is good to hear.

On the CCM concerns -- I think you may have been looking at `main` rather than the PR branch. Everything is implemented end-to-end:

- **`plaintextLength` is plumbed through to Rust**: Both `op_node_cipheriv_set_aad` and `op_node_decipheriv_set_aad` take `plaintext_length: i32` as a third parameter (`lib.rs:537`, `lib.rs:629`) and forward it via `context.set_aad(aad, pt_len)`.

- **CCM variants exist in `cipher.rs`**: `Aes128Ccm`, `Aes192Ccm`, `Aes256Ccm` are in both `Cipher` and `Decipher` enums, with constructor matching for `"aes-{128,192,256}-ccm"`, nonce/tag validation, and a full `AesCcmCipher` struct implementing RFC 3610 (CBC-MAC + CTR mode) with encrypt/decrypt/auth-tag logic.

File: ext/node/polyfills/internal/crypto/cipher.ts [hunk 1]
```diff
@@ -323,6 +414,18 @@ Cipheriv.prototype.update = function (
   }
   _lazyInitCipherDecoder(this, outputEncoding);
 
+  // CCM: buffer data and return empty; actual encryption in final()
+  if (this._isCCM) {
+    const maxSize = ccmMaxMessageSize(this._ccmIvLength);
+    if (this._ccmDataLength + buf.length > maxSize) {
```
Comment (miracatbot) [inline ext/node/polyfills/internal/crypto/cipher.ts]: Problem: This accepts multiple `update()` calls in CCM mode by buffering all chunks and processing them in `final()`, but Node documents/tests CCM as single-shot (`update()` must be called exactly once).

Why here: The CCM-specific `update()` path just accumulates into `_ccmDataLength` / `push_data()` with no guard against a second call.

Impact: That creates a Node-compat divergence in observable behavior for callers that accidentally stream/chunk CCM input. Node rejects that usage; Deno would silently accept it, so compat tests may still miss a real behavioral mismatch.

Ask: Please add a one-shot guard for CCM `update()` (cipher + decipher), or add a targeted Node-compat test if the divergence is intentional.

**Review by miracatbot — COMMENTED** [review]
Overall, the AES-CCM implementation looks coherent and the Node docs/tests alignment is mostly good, but I found one Node-compat behavior gap around CCM chunking.

Recommendation: comment
Risk: moderate

Prioritized findings:
1. CCM currently appears to allow multiple `update()` calls by buffering all chunks until `final()`, while Node documents/tests CCM as single-shot (`update()` exactly once). I left an inline comment on that path because it changes observable compatibility behavior for callers that stream/chunk input.

I didn’t spot an obvious logic bug in the core AES-CCM flow beyond that compatibility concern.

File: ext/node/polyfills/internal/crypto/cipher.ts (hunk 1 above)
Comment (bartlomieju) [reply ext/node/polyfills/internal/crypto/cipher.ts]: Good catch. Added a one-shot guard in f528765 -- both Cipher and Decipher now throw `"message cannot be fragmented across multiple calls to update"` on a second `update()` call in CCM mode, matching Node's behavior.

**Review by miracatbot — APPROVED** [review]
Re-reviewed the updated AES-CCM PR.

The Node-compat gap I called out earlier is addressed now: CCM `update()` is guarded as single-shot on both cipher and decipher, which matches the Node CCM contract much better.

I don’t see a new high-confidence issue in the updated revision.

Recommendation: approve
Risk: moderate

**Review by fibibot — COMMENTED** [review]
Substance walks RFC 3610 conformant on a careful read of `ext/node_crypto/cipher.rs`'s `AesCcmCipher`:

- **B0 format**: flags byte (`(has_aad ? 0x40 : 0) | ((tag_len-2)/2 << 3) | (q-1)`) + nonce + plaintext-length encoded in q bytes. ✓ Per §2.2.
- **A0/Ai counter format**: flags byte (`q-1`, no AAD/MAC bits) + nonce + counter encoded in q bytes. ✓ Per §2.3.
- **CBC-MAC**: B0 → encrypted, then for each AAD chunk and plaintext chunk, XOR into the running state and re-encrypt. ✓
- **Tag generation**: encrypt MAC ⊕ encrypt(A0). ✓
- **Constant-time tag comparison** via XOR-OR diff accumulation. ✓ Standard pattern.
- **JS-side validations** at the right layer (`Cipheriv` in `cipher.ts`): IV length 7-13, authTagLength required for CCM, `ccmMaxMessageSize(ivLength) = 2^(8*q) - 1` upper bound check on plaintext length so the q-byte encoding doesn't silently truncate.

CI fully green: 111/112 success, 1 skip — including the new `parallel/test-crypto-authenticated.js` enable in `tests/node_compat/config.jsonc` (the canonical Node CCM test that exercises NIST CCM test vectors plus invalid-tag rejection).

## Holding COMMENT — escalation: custom hand-rolled CCM

This is a 1108-line PR adding a custom AES-CCM implementation. The implementation looks correct on the read above, but custom AEAD primitives are a category where I'd want a dedicated crypto-aware reviewer's eyes — RFC 3610 is notoriously easy to get subtly wrong (B0 flag encoding, AAD length-prefix branch points, padding, tag-CTR application order).

**Why this isn't using `ccm`?** The Rust `ccm` crate (`Aes128/192/256Ccm`) is a well-tested implementation that wraps the same `aes` crate this PR already depends on. Adopting it would:

- Eliminate ~800 lines of crypto-critical custom code from `node_crypto::cipher::AesCcmCipher`.
- Get NIST-vector test coverage from the `ccm` crate's own test suite for free.
- Match the precedent set by the just-landed sign/verify (#33083), GCM (#33079), and ChaCha20-Poly1305 (#33084) extractions, which use `aes_gcm` / `chacha20poly1305` rather than hand-rolling.

Worth at least a paragraph in the PR body on why `ccm` was rejected if it's a deliberate choice (e.g., a streaming-API or feature-gap incompatibility with Node's CCM contract).

## Smaller observations

1. **AAD length encoding gap** (cipher.rs:`cbc_mac`): handles `len < 0xFF00` (2-byte prefix) and `len >= 0xFF00` (`0xFFFE` + 4-byte length). RFC 3610 §2.2.1 also defines a `0xFFFF` prefix + 8-byte length for AAD lengths 2^32 to 2^64-1. The current code silently misencodes those (the `>> 24`/`>> 16` shifts truncate). Practically unreachable since AAD > 4GB is rare, but worth a `debug_assert!(alen < (1u64 << 32))` or a defensive `Err(_)` for the 64-bit range.
2. **`ctr_process` overflow**: `ctr += 1` on `u64`. After 2^64 blocks (theoretically unreachable on any single message — CCM caps message length to 2^(8*q) blocks where q ≤ 8), this would wrap. Since plaintext-length validation caps the input, this is unreachable in practice — just noting.
3. **`aad: Option<Vec<u8>>` then `is_some_and(|a| !a.is_empty())`**: `setAAD([])` and unset both produce no-AAD CBC-MAC behavior, matching the spec's `has_aad` flag. ✓

## Direction

cc @bartlomieju (and ideally a second reviewer with crypto experience) — happy to flip to APPROVE on a yes/no decision about `ccm` crate adoption. If the conscious choice is to keep the custom implementation, I'd want an explicit cross-reference to NIST SP 800-38C test vectors in the test file (the upstream Node test exercises some but not the full NIST set) to lock in the corner cases.
